doc/zk-integration.md

```diff
@@ -0,0 +1,261 @@
+  the Rollup) were the same, this gadget would not be necessary. Thus, if updating the zkVM circuit is possible in
+  practice, by using the Espresso commitment scheme inside the zkVM gadget, one can remove the need for the _COMMs
+  Equivalence_ gadget.
+- The _zkVM_ gadget is the original gadget of the rollup circuit that proves a correct transition from state
+  `cm_state_vm i` to the next state `cm_state_vm i+1` when applying the transactions represented by the commitment value
+  `cm_txs_rollup`.
+- These three gadgets above return a boolean: true if the verification succeeds and false otherwise.
+- For the circuit to accept, all these gadget outputs must be true, and thus we add an _AND_ gate.
+
+The pseudocode of the rollup contract below shows that in the case we rely on the Espresso light client contract to
+fetch the Espresso state, the only inputs to the function `updateRollupState` are `newVMState`, `commTxsRollup` and
+`snarkProof`.
+
+```solidity
+/// Uses the Espresso light client contract to fetch the last state.
+contract RollupContract1 is RollupContract {
+
+    function updateRollupState(
+        newVMState,
+        commTxsRollup,
+        snarkProof){
+
+        // Escape hatch is activated, switch to default sequencing mode
+        if (isEscapeHatchActivated()){
+            this.updateStateBackupSequencingMode(commTxsRollup);
+        } else { // No escape hatch, use the state of Espresso consensus
+            lightClientState = lcContract.getFinalizedState();
+            newEspressoState = lightClientState.blockCommRoot;
+            blockNumberEspresso = lightClientState.blockHeight;
+            this.updateStateFromEspresso(newEspressoState, blockNumberEspresso, commTxsRollup, newVMState, snarkProof);
+        }
+    }
+}
+```
+
+## Integration 2: Verify Espresso consensus inside the rollup circuit
+
+![image](zk-rollup-circuit.svg) **Figure 3:** Rollup circuit with additional gadgets _Espresso Consensus_, _Collect &
+Filter_, and _COMMs Equivalence_. Private inputs of the circuit are written in bold font.
+
+The circuit depicted in Figure 3 operates as follows:
+
+- The _Espresso Consensus_ gadget checks that the block commitment for Espresso block `BLOCK_NUMBER` is `blk_cm`. To
+  achieve this goal, it is required to obtain the state of the stake table from the previous HotShot epoch. Assuming the
+  rollup state is updated at least once per epoch, the commitment `blk_cm_old` will be computed from such state. Hence,
+  the private input `STAKE_TABLE_ENTRIES`, containing the list of public keys with their respective stake, can be linked
+  to the commitment `blk_cm_old` via the private inputs `STAKE_TABLE_OPENINGS`. Finally, note that the first
+  `blk_cm_old` value needs to be read from the light client contract. Afterward, no dependency on the light client
+  contract is needed.
+- The other gadgets are the same as in Integration 1.
```

Comment:
Should we also briefly explain the `STATE_SIGS` input or is the meaning implied by the above?

Reply:
Good catch, the mention of STATE_SIGS was deleted in a previous update. Fixed in 898eaf54.